### PR TITLE
chore: split nstemplateset controller

### PR DESCRIPTION
**Pull request by MatousJobanek**
Since the NSTemplateSet controller (and mainly the associated test) was getting bigger and bigger and thus also cluttered I decided to split it into several files. As part of this effort, I created additional types to isolate the behavior for easier testing. One of the new types is `apiClient` that contains information for communication through the API and is used in all other types via composition.

* `cluster_resources.go` - takes care of creating, updating, and deleting of cluster resources contained in the respective template. The main entity is `clusterResourcesManager`.
* `namespaces.go` - takes care of creating, updating, and deleting of namespaces and its inner resources. The main entity is `namespacesManager`.
* `status.go` - takes care of updating NSTemplateSet.status. The main entity is `statusManager` that is used via composition in both of the previously mentioned managers.

For all the beforementioned files, there are also associated test files testing the behavior in an isolated way. 

This PR doesn't bring any new functionality, it just moves the code to the new files + introduces some refactoring to make the move smooth and meaningful. It may contain some small fixes of bugs that were discovered during the refactoring. 
On the other hand, the unit tests contain bigger changes since in the new code it verifies the behavior in an isolated way. Some tests were added to improve the coverage, duplicated were removed.

### CODE REVIEW

File: pkg/controller/nstemplateset/nstemplateset_controller.go [hunk 1]
```diff
@@ -452,200 +235,38 @@ Current:
 	return deleted, nil
 }
 
+// fetchNamespaces returns all current namespaces belonging to the given user
+// i.e., labeled with `"toolchain.dev.openshift.com/owner":<username>`
+func fetchNamespaces(client client.Client, username string) ([]corev1.Namespace, error) {
```
Comment (alexeykazakov) [inline pkg/controller/nstemplateset/nstemplateset_controller.go:240]: Do you think we can/should move it to namespaces.go too?

Comment (MatousJobanek) [reply pkg/controller/nstemplateset/nstemplateset_controller.go]: whoops, that was somehow forgotten :man_shrugging:  moved in 34e547a

File: pkg/controller/nstemplateset/cluster_resources.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+			}
+			logger.Info("updated cluster resource")
+			return true, r.setStatusUpdatingIfNotProvisioning(nsTmplSet)
+		}
+	}
+	logger.Info("cluster resources already provisioned")
+	return false, nil
+}
+
+func (r *clusterResourcesManager) delete(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/controller/nstemplateset/cluster_resources.go:104]: added in ce0c8e8

File: pkg/controller/nstemplateset/namespaces.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to update namespace '%s'", nsName)
+	}
+
+	logger.Info("namespace provisioned with all required resources", "tier", nsTmplSet.Spec.TierName, "namespace", tcNamespace)
+
+	// TODO add validation for other objects
+	return nil // nothing changed, no error occurred
+}
+
+func (r *namespacesManager) delete(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/controller/nstemplateset/namespaces.go:156]: added in ce0c8e8

**Pull request merged**